File: ruby/examples/complete_workflow.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Complete Workflow Example for JewelMusic Ruby SDK
#
# This example demonstrates a comprehensive music production workflow:
# - Upload and process audio tracks
# - AI-powered analysis and enhancement
# - Automated transcription and lyrics enhancement
# - AI-assisted composition
# - Release creation and distribution
# - Analytics and monitoring

require_relative '../lib/jewelmusic'

# Initialize with environment variable
api_key = ENV['JEWELMUSIC_API_KEY']

if api_key.nil? || api_key.empty?
  puts "❌ JEWELMUSIC_API_KEY environment variable not set"
  puts "Please set your API key: export JEWELMUSIC_API_KEY=your_key_here"
  exit 1
end

puts "🎵 JewelMusic Ruby SDK - Complete Workflow Example"
puts "==================================================\n\n"
puts "🔑 Using API key: #{api_key[0, 12]}...\n\n"

begin
  # Initialize the SDK
  jewelmusic = JewelMusic::Client.new(api_key: api_key)
  
  # Run complete workflow
  results = run_complete_workflow(jewelmusic)
  
  puts "\n✨ Complete workflow examples completed successfully!"
  print_workflow_summary(results)
  
rescue JewelMusic::Exceptions::ApiError => e
  puts "\n💥 API Error: #{e.message}"
  puts "Status Code: #{e.status_code}"
  puts "Details: #{e.details}" if e.details
  exit 1
rescue StandardError => e
  puts "\n💥 Unexpected Error: #{e.message}"
  exit 1
end

def run_complete_workflow(jewelmusic)
  puts "🎵 JewelMusic Complete Workflow Example"
  puts "======================================"
  
  workflow_results = {}
  
  # Phase 1: Upload and Processing
  track = phase1_upload_and_processing(jewelmusic)
  workflow_results[:track] = track
  
  # Phase 2: AI Analysis
  analysis = phase2_ai_analysis(jewelmusic, track)
  workflow_results[:analysis] = analysis
  
  # Phase 3: Transcription and Lyrics
  transcription = phase3_transcription_and_lyrics(jewelmusic, track, analysis)
  workflow_results[:transcription] = transcription
  
  # Phase 4: AI Composition
  composition = phase4_ai_composition(jewelmusic, track, analysis)
  workflow_results[:composition] = composition
  
  # Phase 5: Release and Distribution
  distribution = phase5_release_and_distribution(jewelmusic, track, analysis)
  workflow_results[:distribution] = distribution
  
  # Phase 6: Analytics Setup
  analytics = phase6_analytics_setup(jewelmusic, distribution)
  workflow_results[:analytics] = analytics
  
  workflow_results
end

def phase1_upload_and_processing(jewelmusic)
  puts "\n📤 Phase 1: Track Upload and Processing"
  puts "--------------------------------------"
  
  # Check for sample audio file
  audio_file = 'sample-audio.mp3'
  
  if File.exist?(audio_file)
    puts "📁 Found sample audio file, uploading..."
    
    # Upload with metadata
    metadata = {
      title: 'Workflow Demo Track',
      artist: 'SDK Demo Artist',
      album: 'Demo Album',
      genre: 'Electronic',
      release_date: '2025-09-01'
    }
    
    track = jewelmusic.tracks.upload(audio_file, 'sample-audio.mp3', metadata)
    
    puts "✅ Track uploaded successfully!"
    puts "Track ID: #{track['id']}"
    puts "Status: #{track['status']}"
    
    # Wait for processing
    puts "\n⏳ Waiting for track processing..."
    processed_track = wait_for_processing(jewelmusic, track)
    puts "✅ Track processed: #{processed_track['status']}"
    
    processed_track
  else
    puts "⚠️  No sample audio file found, using existing track..."
    
    # Get an existing track
    tracks_response = jewelmusic.tracks.list(page: 1, per_page: 1)
    tracks = tracks_response['items']
    
    if tracks.empty?
      raise "No tracks available. Please upload a track first or add sample-audio.mp3"
    end
    
    track = tracks.first
    puts "✅ Using existing track: #{track['title']}"
    
    track
  end
end

def phase2_ai_analysis(jewelmusic, track)
  puts "\n🔍 Phase 2: AI Analysis and Quality Assessment"
  puts "---------------------------------------------"
  
  # Comprehensive analysis
  analysis_options = {
    analysis_types: %w[tempo key structure quality loudness mood],
    detailed_report: true,
    cultural_context: 'global',
    target_platforms: %w[spotify apple-music youtube-music]
  }
  
  analysis = jewelmusic.analysis.analyze_with_options(track['id'], analysis_options)
  
  puts "✅ Analysis completed!"
  puts "📊 Results:"
  
  if analysis['tempo']
    tempo = analysis['tempo']
    puts "   Tempo: #{tempo['bpm'].round(1)} BPM " \
         "(confidence: #{(tempo['confidence'] * 100).round(1)}%)"
  end
  
  if analysis['key']
    key = analysis['key']
    puts "   Key: #{key['key']} #{key['mode']} " \
         "(confidence: #{(key['confidence'] * 100).round(1)}%)"
  end
  
  if analysis['quality']
    quality = analysis['quality']
    puts "   Quality Score: #{(quality['overallScore'] * 100).round(1)}/100"
  end
  
  if analysis['loudness']
    loudness = analysis['loudness']
    puts "   Loudness: #{loudness['lufs'].round(1)} LUFS"
  end
  
  if analysis['mood']
    mood = analysis['mood']
    puts "   Primary Mood: #{mood['primary']} (#{mood['energy']} energy)"
  end
  
  # Quality recommendations
  if analysis['quality']&.dig('recommendations')&.any?
    puts "\n💡 Quality Recommendations:"
    analysis['quality']['recommendations'].each do |rec|
      puts "   - #{rec['type']}: #{rec['suggestion']}"
    end
  end
  
  analysis
end

def phase3_transcription_and_lyrics(jewelmusic, track, analysis)
  puts "\n📝 Phase 3: Transcription and Lyrics Enhancement"
  puts "-----------------------------------------------"
  
  begin
    # Create transcription
    transcription_options = {
      languages: %w[en auto-detect],
      include_timestamps: true,
      word_level_timestamps: true,
      speaker_diarization: true,
      model: 'large'
    }
    
    transcription = jewelmusic.transcription.create_transcription(track['id'], transcription_options)
    
    puts "✅ Transcription completed!"
    puts "Language detected: #{transcription['language']}"
    puts "Confidence: #{(transcription['confidence'] * 100).round(1)}%"
    
    text = transcription['text']
    preview = text.length > 150 ? "#{text[0, 150]}..." : text
    puts "Text preview: #{preview}"
    
    # Enhance lyrics with AI
    enhancement_options = {
      improve_meter: true,
      enhance_rhyming: true,
      adjust_tone: 'professional',
      preserve_original_meaning: true
    }
    
    enhanced_lyrics = jewelmusic.transcription.enhance_lyrics(text, enhancement_options)
    
    puts "✅ Lyrics enhanced!"
    puts "Enhancements applied: #{enhanced_lyrics['enhancements'].join(', ')}"
    
    enhanced_lyrics
    
  rescue StandardError => e
    puts "⚠️  Transcription not available (instrumental or processing failed)"
    puts "Proceeding with AI-generated lyrics instead..."
    
    # Generate AI lyrics based on mood and style
    mood_primary = analysis.dig('mood', 'primary') || 'uplifting'
    tempo_bpm = analysis.dig('tempo', 'bpm')&.to_i || 120
    key_str = if analysis['key']
                "#{analysis['key']['key']} #{analysis['key']['mode']}"
              else
                'C major'
              end
    
    lyrics_options = {
      theme: "#{mood_primary} electronic music",
      genre: track['genre'] || 'electronic',
      language: 'en',
      mood: mood_primary,
      structure: 'verse-chorus-verse-chorus-bridge-chorus',
      tempo: tempo_bpm,
      key: key_str
    }
    
    transcription = jewelmusic.copilot.generate_lyrics(lyrics_options)
    
    puts "✅ AI lyrics generated!"
    puts "Theme: #{transcription['theme']}"
    
    text = transcription['text']
    preview = text.length > 150 ? "#{text[0, 150]}..." : text
    puts "Text preview: #{preview}"
    
    transcription
  end
end

def phase4_ai_composition(jewelmusic, track, analysis)
  puts "\n🤖 Phase 4: AI-Assisted Composition"
  puts "------------------------------------"
  
  # Extract analysis data
  key_name = analysis.dig('key', 'key') || 'C'
  key_mode = analysis.dig('key', 'mode') || 'major'
  tempo_bpm = analysis.dig('tempo', 'bpm')&.to_i || 120
  mood_primary = analysis.dig('mood', 'primary') || 'uplifting'
  mood_energy = analysis.dig('mood', 'energy') || 'medium'
  
  # Generate composition elements concurrently using threads
  compositions = {}
  threads = []
  
  # Generate harmony
  threads << Thread.new do
    harmony_options = {
      style: track['genre'] || 'electronic',
      key: key_name,
      mode: key_mode,
      tempo: tempo_bpm,
      complexity: 'medium',
      voicing: 'modern'
    }
    
    compositions[:harmony] = jewelmusic.copilot.generate_harmony(harmony_options)
  end
  
  # Generate chord progression
  threads << Thread.new do
    chord_options = {
      key: "#{key_name} #{key_mode}",
      style: track['genre'] || 'electronic',
      complexity: 0.5,
      length: 8
    }
    
    compositions[:chord_progression] = jewelmusic.copilot.generate_chord_progression(chord_options)
  end
  
  # Generate arrangement suggestions
  threads << Thread.new do
    arrangement_options = {
      track_id: track['id'],
      genre: track['genre'] || 'electronic',
      mood: mood_primary,
      duration: track['duration'] || 180,
      energy: mood_energy
    }
    
    compositions[:arrangement] = jewelmusic.copilot.suggest_arrangement(arrangement_options)
  end
  
  # Wait for all threads to complete
  threads.each(&:join)
  
  harmony = compositions[:harmony]
  chord_progression = compositions[:chord_progression]
  arrangement = compositions[:arrangement]
  
  puts "✅ AI composition elements generated!"
  puts "🎼 Harmony ID: #{harmony['id']}"
  puts "🎹 Chord progression: #{chord_progression['progression'].join(' - ')}"
  puts "🎛️  Suggested arrangement: #{arrangement['structure'].join(' → ')}"
  
  # Create a style variation
  style_options = {
    source_id: track['id'],
    target_style: 'ambient',
    intensity: 0.6,
    preserve_structure: true,
    preserve_timing: true
  }
  
  style_variation = jewelmusic.copilot.apply_style_transfer(style_options)
  
  puts "✅ Style variation created!"
  puts "🎨 Variation ID: #{style_variation['id']}"
  puts "Applied style: #{style_variation['appliedStyle']}"
  
  {
    harmony: harmony,
    chord_progression: chord_progression,
    arrangement: arrangement,
    style_variation: style_variation
  }
end

def phase5_release_and_distribution(jewelmusic, track, analysis)
  puts "\n📡 Phase 5: Release Creation and Distribution"
  puts "--------------------------------------------"
  
  # Create release
  track_data = {
    track_id: track['id'],
    title: track['title'],
    duration: track['duration'] || 180,
    isrc: generate_mock_isrc
  }
  
  release_data = {
    type: 'single',
    title: track['title'],
    artist: track['artist'],
    release_date: '2025-09-01',
    tracks: [track_data],
    artwork: {
      primary: true,
      type: 'cover'
    }
  }
  
  # Metadata from analysis
  metadata = {
    genre: track['genre'] || 'Electronic',
    explicit: false
  }
  
  metadata[:subgenre] = analysis.dig('style', 'subgenre') if analysis.dig('style', 'subgenre')
  metadata[:mood] = analysis.dig('mood', 'primary') if analysis.dig('mood', 'primary')
  metadata[:tempo] = analysis.dig('tempo', 'bpm') if analysis.dig('tempo', 'bpm')
  
  if analysis['key']
    metadata[:key] = "#{analysis['key']['key']} #{analysis['key']['mode']}"
  end
  
  metadata[:credits] = [
    { role: 'artist', name: track['artist'] },
    { role: 'producer', name: 'AI-Assisted Production' }
  ]
  
  release_data[:metadata] = metadata
  release_data[:territories] = ['worldwide']
  release_data[:platforms] = %w[spotify apple-music youtube-music soundcloud bandcamp]
  
  release = jewelmusic.distribution.create_release(release_data)
  
  puts "✅ Release created!"
  puts "📦 Release ID: #{release['id']}"
  puts "Status: #{release['status']}"
  
  # Validate release before submission
  validation = jewelmusic.distribution.validate_release(release['id'])
  
  puts "\n🔍 Release validation:"
  puts "Valid: #{validation['valid'] ? 'Yes' : 'No'}"
  
  unless validation['valid']
    puts "⚠️  Validation issues:"
    validation['issues'].each do |issue|
      puts "   - #{issue['severity']}: #{issue['message']}"
    end
  end
  
  # Submit for distribution (if valid)
  submission_result = nil
  if validation['valid']
    submission_data = {
      platforms: %w[spotify apple-music],
      scheduled_date: '2025-09-01T00:00:00Z',
      expedited: false
    }
    
    submission_result = jewelmusic.distribution.submit_to_platforms(release['id'], submission_data)
    
    puts "✅ Release submitted for distribution!"
    puts "Submission ID: #{submission_result['id']}"
    puts "Expected processing time: #{submission_result['estimatedProcessingTime']}"
  end
  
  {
    release: release,
    validation: validation,
    submission: submission_result
  }
end

def phase6_analytics_setup(jewelmusic, distribution)
  puts "\n📊 Phase 6: Analytics Setup and Monitoring"
  puts "------------------------------------------"
  
  release = distribution[:release]
  
  # Setup analytics alerts
  alert_data = {
    name: 'Track Performance Alert',
    release_id: release['id'],
    conditions: [
      {
        metric: 'streams',
        operator: 'greater_than',
        threshold: 1000,
        period: 'day'
      },
      {
        metric: 'completion_rate',
        operator: 'less_than',
        threshold: 0.5,
        period: 'day'
      }
    ],
    notifications: {
      email: true,
      webhook: true,
      dashboard: true
    }
  }
  
  alert = jewelmusic.analytics.setup_alert(alert_data)
  
  puts "✅ Analytics alert created!"
  puts "Alert ID: #{alert['id']}"
  
  # Try to get initial analytics (if track has existing data)
  begin
    tracks = release['tracks']
    track_id = tracks.first['trackId']
    
    analytics_query = {
      start_date: '2025-01-01',
      end_date: '2025-01-31',
      metrics: %w[streams listeners completion_rate skip_rate],
      group_by: 'day'
    }
    
    initial_analytics = jewelmusic.analytics.get_track_analytics(track_id, analytics_query)
    
    summary = initial_analytics['summary']
    total_streams = summary['totalStreams']
    
    if total_streams > 0
      puts "📈 Initial analytics:"
      puts "   Total streams: #{summary['totalStreams']}"
      puts "   Unique listeners: #{summary['uniqueListeners']}"
      puts "   Completion rate: #{(summary['completionRate'] * 100).round(1)}%"
    end
  rescue StandardError
    puts "⚠️  No analytics data available yet (new track)"
  end
  
  { alert: alert }
end

def wait_for_processing(jewelmusic, track)
  # Simulate waiting for processing
  track_id = track['id']
  max_attempts = 10
  attempt = 0
  
  while attempt < max_attempts
    current_track = jewelmusic.tracks.get(track_id)
    status = current_track['status']
    
    return current_track if %w[processed ready].include?(status)
    
    raise "Track processing failed" if %w[failed error].include?(status)
    
    attempt += 1
    sleep(2) # Wait 2 seconds
  end
  
  raise "Track processing timeout"
end

def generate_mock_isrc
  # Generate a mock ISRC code
  timestamp = Time.now.to_i
  "US#{timestamp.to_s[-8..-1]}"
end

def print_workflow_summary(results)
  puts "\n✨ Phase 7: Workflow Summary"
  puts "----------------------------"
  
  track = results[:track]
  analysis = results[:analysis]
  composition = results[:composition]
  transcription = results[:transcription]
  distribution = results[:distribution]
  analytics = results[:analytics]
  
  puts "🎉 Workflow completed successfully!"
  puts "\n📋 Summary:"
  
  # Track summary
  puts "Track: #{track['title']} by #{track['artist']} (ID: #{track['id']})"
  
  # Analysis summary
  if analysis&.dig('tempo') && analysis&.dig('key') && analysis&.dig('quality')
    tempo = analysis['tempo']['bpm'].round(1)
    key = "#{analysis['key']['key']} #{analysis['key']['mode']}"
    quality = (analysis['quality']['overallScore'] * 100).round(1)
    puts "Analysis: #{tempo} BPM, #{key}, Quality #{quality}/100"
  end
  
  # Composition summary
  if composition&.dig(:harmony) && composition&.dig(:style_variation)
    harmony_id = composition[:harmony]['id']
    variation_id = composition[:style_variation]['id']
    puts "Composition: Harmony #{harmony_id}, Style variation #{variation_id}"
  end
  
  # Distribution summary
  if distribution&.dig(:release) && distribution&.dig(:validation)
    release = distribution[:release]
    validation = distribution[:validation]
    valid_text = validation['valid'] ? 'Yes' : 'No'
    puts "Distribution: Release #{release['id']} (#{release['status']}), Valid: #{valid_text}"
  end
  
  puts "\n🚀 Next Steps:"
  puts "1. Monitor distribution status for platform approval"
  puts "2. Track analytics and streaming performance"
  puts "3. Use AI insights for future compositions"
  puts "4. Create variations and remixes using style transfer"
  puts "5. Optimize release strategy based on performance data"
end

File: javascript/examples/complete-workflow.js
/**
 * Complete Workflow Example
 * 
 * This example demonstrates a comprehensive music production workflow using JewelMusic:
 * - Upload and process audio tracks
 * - AI-powered analysis and enhancement
 * - Automated transcription and lyrics enhancement
 * - AI-assisted composition
 * - Release creation and distribution
 * - Analytics and monitoring
 */

const { JewelMusic } = require('@jewelmusic/sdk');
const fs = require('fs');
const path = require('path');

async function completeWorkflowExample() {
  const client = new JewelMusic({
    apiKey: process.env.JEWELMUSIC_API_KEY,
    environment: 'sandbox'
  });

  try {
    console.log('🎵 JewelMusic Complete Workflow Example');
    console.log('======================================\n');

    // Phase 1: Track Upload and Processing
    console.log('📤 Phase 1: Track Upload and Processing');
    console.log('--------------------------------------');

    const audioPath = path.join(__dirname, 'sample-audio.mp3');
    let track;

    if (fs.existsSync(audioPath)) {
      console.log('📁 Found sample audio file, uploading...');
      
      const audioFile = fs.readFileSync(audioPath);
      track = await client.tracks.upload(audioFile, {
        title: 'Workflow Demo Track',
        artist: 'SDK Demo Artist',
        album: 'Demo Album',
        genre: 'Electronic',
        releaseDate: '2025-09-01'
      }, {
        onProgress: (progress) => {
          process.stdout.write(`\rUpload progress: ${progress.percentage}%`);
        }
      });

      console.log('\n✅ Track uploaded successfully!');
      console.log('Track ID:', track.id);
      console.log('Status:', track.status);

      // Wait for processing
      console.log('\n⏳ Waiting for track processing...');
      const processedTrack = await client.tracks.waitForProcessing(track.id);
      console.log('✅ Track processed:', processedTrack.status);

    } else {
      console.log('⚠️  No sample audio file found, using existing track...');
      
      // Get an existing track
      const tracks = await client.tracks.list({ page: 1, perPage: 1 });
      if (tracks.items.length === 0) {
        throw new Error('No tracks available for demonstration. Please upload a track first or add sample-audio.mp3');
      }
      
      track = tracks.items[0];
      console.log('✅ Using existing track:', track.title);
    }

    // Phase 2: AI Analysis and Quality Assessment
    console.log('\n🔍 Phase 2: AI Analysis and Quality Assessment');
    console.log('---------------------------------------------');

    // Comprehensive analysis
    const analysis = await client.analysis.uploadTrack(audioPath || null, {
      trackId: track.id,
      analysisTypes: ['tempo', 'key', 'structure', 'quality', 'loudness', 'mood'],
      detailedReport: true,
      culturalContext: 'global',
      targetPlatforms: ['spotify', 'apple-music', 'youtube-music']
    });

    console.log('✅ Analysis completed!');
    console.log('📊 Results:');
    console.log(`   Tempo: ${analysis.tempo.bpm} BPM (confidence: ${analysis.tempo.confidence})`);
    console.log(`   Key: ${analysis.key.key} ${analysis.key.mode} (confidence: ${analysis.key.confidence})`);
    console.log(`   Quality Score: ${analysis.quality.overallScore}/1.0`);
    console.log(`   Loudness: ${analysis.loudness.lufs} LUFS`);
    console.log(`   Primary Mood: ${analysis.mood.primary} (${analysis.mood.energy} energy)`);

    // Quality recommendations
    if (analysis.quality.recommendations) {
      console.log('\n💡 Quality Recommendations:');
      analysis.quality.recommendations.forEach(rec => {
        console.log(`   - ${rec.type}: ${rec.suggestion}`);
      });
    }

    // Phase 3: Transcription and Lyrics Enhancement
    console.log('\n📝 Phase 3: Transcription and Lyrics Enhancement');
    console.log('-----------------------------------------------');

    let transcription;
    try {
      // Create transcription
      transcription = await client.transcription.create(track.id, {
        languages: ['en', 'auto-detect'],
        includeTimestamps: true,
        wordLevelTimestamps: true,
        speakerDiarization: true,
        model: 'large'
      });

      console.log('✅ Transcription completed!');
      console.log('Language detected:', transcription.language);
      console.log('Confidence:', transcription.confidence);
      console.log('Text preview:', transcription.text.substring(0, 150) + '...');

      // Enhance lyrics with AI
      const enhancedLyrics = await client.transcription.enhanceLyrics(
        transcription.text,
        {
          improveMeter: true,
          enhanceRhyming: true,
          adjustTone: 'professional',
          preserveOriginalMeaning: true
        }
      );

      console.log('✅ Lyrics enhanced!');
      console.log('Enhancements applied:', enhancedLyrics.enhancements.join(', '));

    } catch (error) {
      console.log('⚠️  Transcription not available for this track (instrumental or processing failed)');
      console.log('Proceeding with AI-generated lyrics instead...');

      // Generate AI lyrics based on the mood and style
      transcription = await client.copilot.generateLyrics({
        theme: `${analysis.mood.primary} electronic music`,
        genre: track.genre || 'electronic',
        language: 'en',
        mood: analysis.mood.primary,
        structure: 'verse-chorus-verse-chorus-bridge-chorus',
        tempo: analysis.tempo.bpm,
        key: `${analysis.key.key} ${analysis.key.mode}`
      });

      console.log('✅ AI lyrics generated!');
      console.log('Theme:', transcription.theme);
      console.log('Text preview:', transcription.text.substring(0, 150) + '...');
    }

    // Phase 4: AI-Assisted Composition
    console.log('\n🤖 Phase 4: AI-Assisted Composition');
    console.log('------------------------------------');

    // Generate complementary elements
    const compositionElements = await Promise.all([
      // Generate harmony
      client.copilot.generateHarmony({
        style: track.genre || 'electronic',
        key: analysis.key.key,
        mode: analysis.key.mode,
        tempo: analysis.tempo.bpm,
        complexity: 'medium',
        voicing: 'modern'
      }),

      // Generate chord progression
      client.copilot.chordProgression({
        key: analysis.key.key,
        mode: analysis.key.mode,
        style: track.genre || 'electronic',
        complexity: 'medium',
        length: 8
      }),

      // Generate arrangement suggestions
      client.copilot.suggestArrangement({
        trackId: track.id,
        genre: track.genre || 'electronic',
        mood: analysis.mood.primary,
        duration: track.duration,
        energy: analysis.mood.energy
      })
    ]);

    const [harmony, chordProgression, arrangement] = compositionElements;

    console.log('✅ AI composition elements generated!');
    console.log('🎼 Harmony ID:', harmony.id);
    console.log('🎹 Chord progression:', chordProgression.progression.join(' - '));
    console.log('🎛️  Suggested arrangement:', arrangement.structure.join(' → '));

    // Create a remix or variation
    const styleVariation = await client.copilot.styleTransfer({
      sourceId: track.id,
      targetStyle: 'ambient',
      intensity: 0.6,
      preserveStructure: true,
      preserveTiming: true
    });

    console.log('✅ Style variation created!');
    console.log('🎨 Variation ID:', styleVariation.id);
    console.log('Applied style:', styleVariation.appliedStyle);

    // Phase 5: Release Creation and Distribution
    console.log('\n📡 Phase 5: Release Creation and Distribution');
    console.log('--------------------------------------------');

    // Create release
    const release = await client.distribution.createRelease({
      type: 'single',
      title: track.title,
      artist: track.artist,
      releaseDate: '2025-09-01',
      tracks: [{
        trackId: track.id,
        title: track.title,
        duration: track.duration,
        isrc: `US${Date.now().toString().slice(-8)}` // Generate mock ISRC
      }],
      artwork: {
        primary: true,
        type: 'cover',
        // In real scenario, you'd upload artwork
      },
      metadata: {
        genre: track.genre,
        subgenre: analysis.style?.subgenre,
        mood: analysis.mood.primary,
        tempo: analysis.tempo.bpm,
        key: `${analysis.key.key} ${analysis.key.mode}`,
        explicit: false,
        credits: [
          { role: 'artist', name: track.artist },
          { role: 'producer', name: 'AI-Assisted Production' }
        ]
      },
      territories: ['worldwide'],
      platforms: ['spotify', 'apple-music', 'youtube-music', 'soundcloud', 'bandcamp']
    });

    console.log('✅ Release created!');
    console.log('📦 Release ID:', release.id);
    console.log('Status:', release.status);

    // Validate release before submission
    const validation = await client.distribution.validateRelease(release.id);
    console.log('\n🔍 Release validation:');
    console.log('Valid:', validation.valid);
    
    if (!validation.valid) {
      console.log('⚠️  Validation issues:');
      validation.issues.forEach(issue => {
        console.log(`   - ${issue.severity}: ${issue.message}`);
      });
    }

    // Submit for distribution (if valid)
    if (validation.valid) {
      const submission = await client.distribution.submitToPlatforms(release.id, {
        platforms: ['spotify', 'apple-music'],
        scheduledDate: '2025-09-01T00:00:00Z',
        expedited: false
      });

      console.log('✅ Release submitted for distribution!');
      console.log('Submission ID:', submission.id);
      console.log('Expected processing time:', submission.estimatedProcessingTime);
    }

    // Phase 6: Analytics Setup and Monitoring
    console.log('\n📊 Phase 6: Analytics Setup and Monitoring');
    console.log('------------------------------------------');

    // Setup analytics alerts
    const alert = await client.analytics.setupAlert({
      name: 'Track Performance Alert',
      releaseId: release.id,
      conditions: [
        {
          metric: 'streams',
          operator: 'greater_than',
          threshold: 1000,
          period: 'day'
        },
        {
          metric: 'completion_rate',
          operator: 'less_than',
          threshold: 0.5,
          period: 'day'
        }
      ],
      notifications: {
        email: true,
        webhook: true,
        dashboard: true
      }
    });

    console.log('✅ Analytics alert created!');
    console.log('Alert ID:', alert.id);

    // Get initial analytics (if track has existing data)
    try {
      const initialAnalytics = await client.analytics.getTrackAnalytics(track.id, {
        startDate: '2025-01-01',
        endDate: '2025-01-31',
        metrics: ['streams', 'listeners', 'completion_rate', 'skip_rate'],
        groupBy: 'day'
      });

      if (initialAnalytics.summary.totalStreams > 0) {
        console.log('📈 Initial analytics:');
        console.log('   Total streams:', initialAnalytics.summary.totalStreams);
        console.log('   Unique listeners:', initialAnalytics.summary.uniqueListeners);
        console.log('   Completion rate:', (initialAnalytics.summary.completionRate * 100).toFixed(1) + '%');
      }
    } catch (error) {
      console.log('⚠️  No analytics data available yet (new track)');
    }

    // Phase 7: Workflow Summary and Next Steps
    console.log('\n✨ Phase 7: Workflow Summary');
    console.log('----------------------------');

    const workflowSummary = {
      track: {
        id: track.id,
        title: track.title,
        artist: track.artist,
        duration: track.duration,
        status: track.status
      },
      analysis: {
        tempo: analysis.tempo.bpm,
        key: `${analysis.key.key} ${analysis.key.mode}`,
        quality: analysis.quality.overallScore,
        mood: analysis.mood.primary
      },
      composition: {
        harmonyId: harmony.id,
        chordProgression: chordProgression.progression.slice(0, 4).join(' - '),
        styleVariationId: styleVariation.id
      },
      transcription: {
        id: transcription.id,
        language: transcription.language,
        hasEnhancements: !!transcription.enhancements
      },
      release: {
        id: release.id,
        status: release.status,
        platforms: release.platforms.length,
        valid: validation.valid
      },
      analytics: {
        alertId: alert.id,
        monitoring: true
      }
    };

    console.log('🎉 Workflow completed successfully!');
    console.log('\n📋 Summary:');
    console.log(JSON.stringify(workflowSummary, null, 2));

    console.log('\n🚀 Next Steps:');
    console.log('1. Monitor distribution status for platform approval');
    console.log('2. Track analytics and streaming performance');
    console.log('3. Use AI insights for future compositions');
    console.log('4. Create variations and remixes using style transfer');
    console.log('5. Optimize release strategy based on performance data');

    return workflowSummary;

  } catch (error) {
    console.error('\n❌ Workflow error:', error.message);
    
    if (error.name === 'AuthenticationError') {
      console.log('💡 Make sure your API key is set: export JEWELMUSIC_API_KEY=your_key_here');
    } else if (error.name === 'ValidationError') {
      console.log('💡 Check input data:', error.details);
    } else if (error.name === 'RateLimitError') {
      console.log('💡 API rate limit exceeded. Try again in', error.retryAfter, 'seconds');
    }
    
    throw error;
  }
}

// Batch processing example
async function batchProcessingExample() {
  const client = new JewelMusic({
    apiKey: process.env.JEWELMUSIC_API_KEY,
    environment: 'sandbox'
  });

  console.log('\n⚡ Batch Processing Example');
  console.log('==========================');

  try {
    // Get multiple tracks for batch processing
    const tracks = await client.tracks.list({ page: 1, perPage: 5 });
    
    if (tracks.items.length === 0) {
      console.log('⚠️  No tracks available for batch processing');
      return;
    }

    console.log(`📁 Processing ${tracks.items.length} tracks in batch...`);

    // Batch analyze all tracks
    const batchAnalyses = await Promise.all(
      tracks.items.map(async (track, index) => {
        try {
          console.log(`🔍 Analyzing track ${index + 1}/${tracks.items.length}: ${track.title}`);
          
          const analysis = await client.analysis.analyzeTrack(track.id, {
            analysisTypes: ['tempo', 'key', 'mood'],
            detailedReport: false
          });

          return {
            trackId: track.id,
            title: track.title,
            analysis,
            success: true
          };
        } catch (error) {
          return {
            trackId: track.id,
            title: track.title,
            error: error.message,
            success: false
          };
        }
      })
    );

    // Summary of batch processing
    const successful = batchAnalyses.filter(r => r.success);
    const failed = batchAnalyses.filter(r => !r.success);

    console.log('\n📊 Batch Processing Results:');
    console.log(`✅ Successful: ${successful.length}`);
    console.log(`❌ Failed: ${failed.length}`);

    if (successful.length > 0) {
      console.log('\n🎵 Analysis Results:');
      successful.forEach(result => {
        console.log(`- ${result.title}:`);
        console.log(`  Tempo: ${result.analysis.tempo.bpm} BPM`);
        console.log(`  Key: ${result.analysis.key.key} ${result.analysis.key.mode}`);
        console.log(`  Mood: ${result.analysis.mood.primary}`);
      });
    }

    if (failed.length > 0) {
      console.log('\n❌ Failed Analyses:');
      failed.forEach(result => {
        console.log(`- ${result.title}: ${result.error}`);
      });
    }

    return batchAnalyses;

  } catch (error) {
    console.error('❌ Batch processing error:', error.message);
    throw error;
  }
}

// Performance monitoring example
async function performanceMonitoringExample() {
  const client = new JewelMusic({
    apiKey: process.env.JEWELMUSIC_API_KEY,
    environment: 'sandbox'
  });

  console.log('\n📈 Performance Monitoring Example');
  console.log('=================================');

  try {
    // Get streaming analytics
    const streamingData = await client.analytics.getStreams({
      startDate: '2025-01-01',
      endDate: '2025-01-31',
      groupBy: 'day',
      platforms: ['spotify', 'apple-music'],
      metrics: ['streams', 'listeners', 'revenue']
    });

    console.log('📊 Streaming Performance:');
    console.log(`Total Streams: ${streamingData.summary.totalStreams}`);
    console.log(`Unique Listeners: ${streamingData.summary.uniqueListeners}`);
    console.log(`Revenue: $${streamingData.summary.totalRevenue}`);

    // Get audience demographics
    const audience = await client.analytics.getAudienceAnalytics({
      startDate: '2025-01-01',
      endDate: '2025-01-31',
      groupBy: 'country'
    });

    console.log('\n🌍 Top Countries:');
    audience.demographics.countries.slice(0, 5).forEach(country => {
      console.log(`- ${country.name}: ${country.streams} streams (${country.percentage}%)`);
    });

    // Get real-time analytics
    const realtime = await client.analytics.getRealtimeAnalytics({
      period: 'last_24_hours',
      updateInterval: 300000 // 5 minutes
    });

    console.log('\n⚡ Real-time Data (Last 24h):');
    console.log(`Current streams: ${realtime.currentStreams}`);
    console.log(`Active listeners: ${realtime.activeListeners}`);
    console.log(`Trending platforms: ${realtime.trendingPlatforms.join(', ')}`);

  } catch (error) {
    console.log('⚠️  Analytics data not available or insufficient data');
  }
}

// Run the complete workflow
if (require.main === module) {
  console.log('🎵 JewelMusic Complete Workflow Example\n');
  
  // Check for API key
  if (!process.env.JEWELMUSIC_API_KEY) {
    console.error('❌ JEWELMUSIC_API_KEY environment variable not set');
    console.log('Please set your API key: export JEWELMUSIC_API_KEY=your_key_here');
    process.exit(1);
  }

  completeWorkflowExample()
    .then(() => {
      console.log('\n' + '='.repeat(50));
      return batchProcessingExample();
    })
    .then(() => {
      console.log('\n' + '='.repeat(50));
      return performanceMonitoringExample();
    })
    .then(() => {
      console.log('\n✨ All workflow examples completed successfully!');
      console.log('\nFor more examples, check out:');
      console.log('- basic-usage.js - Basic SDK functionality');
      console.log('- ai-generation.js - AI music generation');
      console.log('- webhook-handling.js - Real-time event handling');
    })
    .catch((error) => {
      console.error('\n💥 Workflow examples failed:', error.message);
      process.exit(1);
    });
}

module.exports = {
  completeWorkflowExample,
  batchProcessingExample,
  performanceMonitoringExample
};